### PR TITLE
Add EXP-3200 [v118] Add Initial config for experiment integration testing.

### PR DESCRIPTION
**Pull request by jrbenny35**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/EXP-3200)

## :bulb: Description
This adds a configuration similar to the syncintegration tests that will allow us to test experiments with automation. These will run on their own but should exist here.

### CODE REVIEW

File: Tests/XCUITests/ExperimentIntegrationTests.swift [hunk 1]
```diff
@@ -0,0 +1,69 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import XCTest
+
+final class ExperimentIntegrationTests: BaseTestCase {
```
Comment (lmarceau) [inline Tests/XCUITests/ExperimentIntegrationTests.swift:7]: Hello! Looking into this code it seems there will be some Swiftlint rules failing. You can check that by running it locally:
1. [Installing](https://formulae.brew.sh/formula/swiftlint) Swiftlint with Brew
2. At root of the project run `swiftlint`, and it will tell you if there's anything to fix (in which file, what line, etc).

Comment (jrbenny35) [reply Tests/XCUITests/ExperimentIntegrationTests.swift:7]: Updated based on this

File: Tests/XCUITests/ExperimentIntegrationTests.swift [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+        // open experiments menu
+        let experiments = app.tables.cells.containing(NSPredicate(format: "label CONTAINS 'Experiments'"))
+        experiments.element.tap()
+
+        // match json experiment name
+        let jsonString = ProcessInfo.processInfo.environment["EXP_NAME"]!
+        let experiment = app.tables.cells.containing(NSPredicate(format: "label CONTAINS[c] '\(jsonString)'"))
+        if experiment.count > 0 {
```
Comment (lmarceau) [inline Tests/XCUITests/ExperimentIntegrationTests.swift:44]: Swiftlint failed due to ```Prefer checking `isEmpty` over comparing `count` to zero```
```suggestion
        if !experiment.isEmpty {
```

File: Tests/XCUITests/ExperimentIntegrationTests.swift [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+    func testMessageNavigatesCorrectly() throws {
+        let surveyPredicate = NSPredicate(format: "identifier CONTAINS 'takeSurveyButton'")
+        let surveyLink = app.buttons.matching(surveyPredicate)
+        let surveyURLPredicate = NSPredicate(format: "identifier CONTAINS 'url'")
+
+        surveyLink.element.tap()
+        waitUntilPageLoad()
+        let surveyURL = app.textFields.matching(surveyURLPredicate)
+        if surveyURL.count > 0 {
```
Comment (lmarceau) [inline Tests/XCUITests/ExperimentIntegrationTests.swift:61]: same
```suggestion
        if !surveyURL.isEmpty {
```

File: Tests/XCUITests/ExperimentIntegrationTests.swift [hunk 4]
```diff
@@ -0,0 +1,67 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import XCTest
+
+final class ExperimentIntegrationTests: BaseTestCase {
+
```
Comment (lmarceau) [inline Tests/XCUITests/ExperimentIntegrationTests.swift:8]: Removing empty line due to swiftlint `Don't include vertical whitespace (empty line) after opening braces`
```suggestion
```

Comment (jrbenny35) [reply Tests/XCUITests/ExperimentIntegrationTests.swift:8]: Do you have custom settings for swiftlint? Mine didn't say any of these

Comment (lmarceau) [reply Tests/XCUITests/ExperimentIntegrationTests.swift:8]: I didn't pull the PR, those are from Bitrise. We don't run with anything special, just need to make sure to be at the root of the project + have latest version of Swiftlint.

**Review by lmarceau — APPROVED** [review]
I'll let @isabelrios do the final approval!

**Comment by jrbenny35** [comment]
Thanks for that title update!

File: Tests/Fennec_Enterprise_XCUITests.xctestplan [hunk 5]
```diff
@@ -24,7 +24,8 @@
       "containerPath" : "container:Client.xcodeproj",
       "identifier" : "F84B21BD1A090F8100AAB793",
       "name" : "Client"
-    }
```
Comment (isabelrios) [inline Tests/Fennec_Enterprise_XCUITests.xctestplan:27]: Do we need changes in this test plan?

File: Tests/XCUITests/ExperimentIntegrationTests.swift [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    func testMessageNavigatesCorrectly() throws {
+        let surveyPredicate = NSPredicate(format: "identifier CONTAINS 'takeSurveyButton'")
+        let surveyLink = app.buttons.matching(surveyPredicate)
+        let surveyURLPredicate = NSPredicate(format: "identifier CONTAINS 'url'")
+
+        surveyLink.element.tap()
+        waitUntilPageLoad()
+        let surveyURL = app.textFields.matching(surveyURLPredicate)
```
Comment (isabelrios) [inline Tests/XCUITests/ExperimentIntegrationTests.swift:56]: we have a `waitForElement` helper method that may be useful for you as it wait for an element to be present before doing anything...

File: ExperimentIntegrationTests/xcrun.py [hunk 7]
```diff
@@ -0,0 +1,21 @@
+import logging
+import subprocess
+
+logging.getLogger(__name__).addHandler(logging.NullHandler())
+
+
+class XCRun(object):
```
Comment (isabelrios) [inline ExperimentIntegrationTests/xcrun.py:7]: could we use the existing xcrun.py file in the SyncIntegration tests? if so, we could move it to a helper or tooling folder...

File: ExperimentIntegrationTests/xcodebuild.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+import os
+import subprocess
+
+from .xcrun import XCRun
+
+here = os.path.dirname(__file__)
+logging.getLogger(__name__).addHandler(logging.NullHandler())
+
+
+class XCodeBuild(object):
```
Comment (isabelrios) [inline ExperimentIntegrationTests/xcodebuild.py:12]: Same comment as below for the xcrun.py file

File: Tests/XCUITests/ExperimentIntegrationTests.swift [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+        }
+        super.setUpScreenGraph()
+    }
+
+    override func tearDownWithError() throws {
+        // Put teardown code here. This method is called after the invocation of each test method in the class.
+    }
+
+    func testVerifyExperimentEnrolled() throws {
+        let toolbarClosePredicate = NSPredicate(format: "label CONTAINS 'Close'")
```
[the remark this answers is not in the document]
Comment (jrbenny35) [reply Tests/XCUITests/ExperimentIntegrationTests.swift:23]: This would be perfect if it also returned the element

File: Tests/Fennec_Enterprise_XCUITests.xctestplan (hunk 5 above)
Comment (jrbenny35) [reply Tests/Fennec_Enterprise_XCUITests.xctestplan:27]: That is a weird change 😅

File: Client.xcodeproj/xcshareddata/xcschemes/Fennec_Enterprise_XCUITests.xcscheme [hunk 10]
```diff
@@ -218,6 +218,9 @@
          <TestPlanReference
             reference = "container:Tests/Smoketest4.xctestplan">
          </TestPlanReference>
+         <TestPlanReference
+            reference = "container:ExperimentIntegrationTests.xctestplan">
```
Comment (isabelrios) [inline Client.xcodeproj/xcshareddata/xcschemes/Fennec_Enterprise_XCUITests.xcscheme:222]: I should have mentioned this before, but our plan in the future is to get rid of this Fennec_Enterprise_XCUITests scheme, so if you can please add this new test plan to the Fennec configuration, similar to the Smoktest plans that we have

File: ExperimentIntegrationTests.xctestplan [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+      "name" : "Configuration 1",
+      "options" : {
+
+      }
+    }
+  ],
+  "defaultOptions" : {
+    "testTimeoutsEnabled" : true
+  },
+  "testTargets" : [
```
Comment (isabelrios) [inline ExperimentIntegrationTests.xctestplan:14]: The test plan appears empty for me. Not sure if I have to clean an re-build everything to see it out of the box. What I did to finally see the test suite in the test plan is to import it from this view:
Screenshot 2023-07-26 at 12 05 04
There is a plus button at the end and you can select your ExperimentIntegrationTests.swift file. 
If it is there for you, then discard my comment.

**Comment by isabelrios** [comment]
@jrbenny35 hope you don't mind, I have converted the PR to draft PR because we still need to work on it a bit

**Comment by isabelrios** [comment]
@jrbenny35 another comment I just realized about is to please move the ExperimentIntegrationTests into the Tests folder. I want to do that eventually with the SyncIntegrationTests too

File: Tests/ExperimentIntegrationTests.xctestplan [hunk 12]
```diff
@@ -0,0 +1,80 @@
+{
```
Comment (isabelrios) [inline Tests/ExperimentIntegrationTests.xctestplan:1]: Why do you have two test plans? are both needed?

File: Tests/XCUITests/ExperimentIntegrationTests.swift [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+        // open experiments menu
+        let experiments = app.tables.cells.containing(NSPredicate(format: "label CONTAINS 'Experiments'"))
+        experiments.element.tap()
+
+        // match json experiment name
+        let experiment = app.tables.cells.containing(NSPredicate(format: "label CONTAINS 'Viewpoint'"))
+        XCTAssertNotNil(experiment)
+    }
+
+    func testMessageNavigatesCorrectly() throws {
```
Comment (isabelrios) [inline Tests/XCUITests/ExperimentIntegrationTests.swift:45]: Question: Even though there are two tests, there was only 1 Passed test result at the end, is that expected? maybe one depend on the other... but I would have expected 2 Passed as there are two tests..

File: Tests/ExperimentIntegrationTests.xctestplan (hunk 12 above)
Comment (isabelrios) [reply Tests/ExperimentIntegrationTests.xctestplan:1]: Also related to test plans, I can't see them correctly on Xcode, even after cleaning and building... do you see them correctly on this (or another view?). It is not listed here, but I can still see it from Fennec -> Edit scheme, however, as you can see there, the test suite `ExperimentIntegrationTests` is not available for the test plan...

Screenshot 2023-07-28 at 12 03 18

File: Tests/XCUITests/ExperimentIntegrationTests.swift (hunk 13 above)
Comment (jrbenny35) [reply Tests/XCUITests/ExperimentIntegrationTests.swift:45]: Well for pytest, it counts as 1

File: Tests/ExperimentIntegrationTests.xctestplan (hunk 12 above)
Comment (jrbenny35) [reply Tests/ExperimentIntegrationTests.xctestplan:1]: Ill remove that older test plan

Comment (jrbenny35) [reply Tests/ExperimentIntegrationTests.xctestplan:1]: Not sure whats going on, I don't know how to fix this 😞

Comment (isabelrios) [reply Tests/ExperimentIntegrationTests.xctestplan:1]: I explained that in a previous comment above: https://github.com/mozilla-mobile/firefox-ios/pull/15574#discussion_r1274746314

Comment (isabelrios) [reply Tests/ExperimentIntegrationTests.xctestplan:1]: oh .. let me try again cleaning my environment and cloning your branch

Comment (isabelrios) [reply Tests/ExperimentIntegrationTests.xctestplan:1]: I think the files are created correctly but we are missing a correct reference between them so that they appear on my side. Let's see if @clarmso can see them correctly though before looking at a solution

File: Tests/ExperimentIntegrationTests/test_integration.py [hunk 14]
```diff
@@ -0,0 +1,8 @@
+def test_testing(xcodebuild, setup_experiment, start_app):
```
Comment (clarmso) [inline Tests/ExperimentIntegrationTests/test_integration.py:1]: Where should we put the `.py` files? The `Tests/` folder contains only `.swift` files so far. The `SyncIntegrationTests` are in the root folder.

Comment (jrbenny35) [reply Tests/ExperimentIntegrationTests/test_integration.py:1]: I was told to move the folder here, see [here](https://github.com/mozilla-mobile/firefox-ios/pull/15574#issuecomment-1653891885)

Comment (isabelrios) [reply Tests/ExperimentIntegrationTests/test_integration.py:1]: This looks good to me @clarmso The struture is the same as the SyncIntegrationTests folder, we have the new folder ExperimentIntegrationTests and the .py files in it.
At some point we should move the SyncIntegrationTests folder into Tests so that all test folder are in the same directory.

**Comment by isabelrios** [comment]
@jrbenny35 as I commented on slack, you have to be sure that the new tests created in your test plan are only enabled in your test plan, not in the others. As you can see in the screenshot, the test suite is enabled for Smoketest2, and the others Fennec's test plans. 
Screenshot 2023-08-07 at 12 34 02


You should right click on the test suite and select disable for each test plan except for yours.

Screenshot 2023-08-07 at 12 45 32

**Review by isabelrios — CHANGES_REQUESTED** [review]
After latest changes this is looking really good, there are however a few things to fix before merging. The existing conflicts and the comment about the new test suite being enabled in all test plans

**Comment by clarmso** [comment]
The `ExperimentIntegrationTests` are enabled on Smoketest2, Smoketest3, Smoketest4 and SmokeXCUITests on my end as well. 😞

**Comment by isabelrios** [comment]
Hi @jrbenny35, sorry for the delay in my review. I was finally able to build and confirm that schemes and test plans are all correct. 
I will approve this PR but before landing there are still two small things:
- Fix the conflict with spm pacakage
- Fix a swiftlint error that is happening:
`ExperimentIntegrationTests.swift:17:47: warning: Overridden methods call super Violation: Method 'tearDownWithError()' should call to super function (overridden_super_call)`

Once you have that fixed I think we are ready to go! thanks!

File: SyncIntegrationTests/xcodebuild.py [hunk 15]
```diff
@@ -16,11 +16,33 @@ class XCodeBuild(object):
...
-    def __init__(self, log):
+    def __init__(self, log, **kwargs):
+        self.scheme = kwargs.get("scheme", self.scheme)
+        self.testPlan = kwargs.get("test_plan", self.testPlan)
         self.log = log
 
-    def test(self, identifier):
-        self.xcrun.erase()
+    def install(self):
+        command = "find ~/Library/Developer/Xcode/DerivedData/Client-*/Build/Products/Fennec-* -type d -iname 'Client.app'"
```
Comment (clarmso) [inline SyncIntegrationTests/xcodebuild.py:25]: This path is the one I usually use. The path for `DerivedData` may differ in a CI environment. Where do the experiments run?

Comment (jrbenny35) [reply SyncIntegrationTests/xcodebuild.py:25]: Hmm, this will run on a Jenkins box, so we might have to update this later

**Review by isabelrios — APPROVED** [review]
Once Bitrise is green this is ready to go! Thanks @jrbenny35

File: SyncIntegrationTests/xcrun.py [hunk 16]
```diff
@@ -13,6 +13,12 @@ def _run(self, *args):
         self.logger.info('Running: {}'.format(' '.join(args)))
         subprocess.check_call(args)
 
+    def boot(self, device='iPhone 14'):
```
Comment (isabelrios) [inline SyncIntegrationTests/xcrun.py:16]: @clarmso please once this lands verify that syncInt tests keep working
